### PR TITLE
Add script to generate Xen unified kernel images

### PR DESCRIPTION
**Pull request by DemiMarie**
This is a prerequisite for secure boot.

Part of https://github.com/QubesOS/qubes-issues/issues/8206

### CODE REVIEW

File: uki-generate [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        f"--change-section-vma=.kernel={kernel_vma}",
+        f"--add-section=.ramdisk={initramfs}",
+        f"--change-section-vma=.ramdisk={initramfs_vma}",
+        "--",
+        hyp,
+        out,
+    ]
+    subprocess.check_call(cmdline, stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL)
+if __name__ == '__main__':
+    main(sys.argv)
```
Comment (fepitre) [inline uki-generate:43]: Can you improve exception handling? Having full trackback may not be nice.

File: uki-generate [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/python3 -I
+import os
+import rpm
+import subprocess
+import sys
+from typing import List, Tuple
+
+def main(args):
+    alignment_mask = (1 << 21) - 1
+    if len(args) != 7 || args[1] != '--':
```
Comment (fepitre) [inline uki-generate:10]:
```suggestion
    if len(args) != 7 or args[1] != '--':
```

**Review by fepitre — CHANGES_REQUESTED** [review]

File: uki-generate [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        f"--change-section-vma=.ramdisk={initramfs_vma}",
+        "--",
+        hyp,
+        out,
+    ]
+    subprocess.check_call(cmdline, stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL)
+if __name__ == '__main__':
+    try:
+        main(sys.argv)
+    except subprocess.CalledProcessError as e:
```
Comment (fepitre) [inline uki-generate:45]:
```suggestion
    except (subprocess.CalledProcessError, ValueError) as e:
        print(str(e))
```

**Review by fepitre — CHANGES_REQUESTED** [review]

File: uki-generate (hunk 3 above)
Comment (DemiMarie) [reply uki-generate:45]: That should not happen.

Comment (DemiMarie) [reply uki-generate:45]: That should not happen.

Comment (fepitre) [reply uki-generate:45]: You raise ValueError in the code so it can happen and we need the error message.

Comment (DemiMarie) [reply uki-generate:45]: `ValueError` is thrown if one of the file sizes is nearly 2<sup>64</sup> bytes.  This is absolutely absurd and I don’t think we need to handle this situation.

Comment (marmarek) [reply uki-generate:45]: If something really shouldn't happen, make it an assert. Other exceptions should be properly handled.

**Comment by marmarek** [comment]
@DemiMarie please add this script to packaging, makefile etc.

File: uki-generate [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import os
+import rpm
+import subprocess
+import sys
+from typing import List, Tuple
+
+def main(args):
+    alignment_mask = (1 << 21) - 1
+    if len(args) != 7 or args[1] != '--':
+        print(f"Usage: uki-generate -- HYPERVISOR CONFIG KERNEL INITRAMFS OUTPUT", file=sys.stderr)
```
Comment (marmarek) [inline uki-generate:11]: Please drop requirement of `--` (here, and generally in future tools you create). In this case, the tool doesn't even support any kind of options!

**Review by marmarek — CHANGES_REQUESTED** [review]

File: uki-generate (hunk 4 above)
Comment (DemiMarie) [reply uki-generate:11]: Ack.

File: uki-generate [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import subprocess
+import sys
+from typing import List, Tuple
+
+def main(args):
+    alignment_mask = (1 << 21) - 1
+    if len(args) != 6:
+        print(f"Usage: uki-generate HYPERVISOR CONFIG KERNEL INITRAMFS OUTPUT", file=sys.stderr)
+        sys.exit(1)
+    _, _, hyp, cfg, kern, initramfs, out = args
```
Comment (fepitre) [inline uki-generate:13]: Too much values to unpack.

**Review by fepitre — CHANGES_REQUESTED** [review]

File: uki-generate [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        out = './' + out
+    output = subprocess.check_output([
+        "objdump",
+        "-WE", # do not use debuginfod
+        "--section-headers",
+        "--",
+        hyp,
+    ])
+    max_vma = 0
+    print(output.decode("UTF-8", 'surrogateescape'))
```
Comment (marmarek) [inline uki-generate:27]: This is leftover debug print, right?

Comment (DemiMarie) [reply uki-generate:27]: Yup!

**Pull request merged**